File: src/common/scoring.py
"""
スコアリング計算関数モジュール

このモジュールは、NagoTimeのコンテキスト対応レコメンドシステムで使用される
各種スコアリング関数を提供する。

Requirements: 4.1〜4.10
Design: セクション6（レコメンドエンジン設計）
"""

from typing import List, Dict, Any, Optional
from dataclasses import dataclass
import math


# デフォルト重み設定
DEFAULT_WEIGHTS = {
    'weather': 0.30,
    'timeslot': 0.25,
    'distance': 0.30,
    'likes': 0.15
}

# 時間帯の隣接関係マッピング
TIMESLOT_ADJACENT = {
    'MORNING': ['NIGHT', 'AFTERNOON'],      # 5-9時: 隣接は4時以前と10時以降
    'AFTERNOON': ['MORNING', 'EVENING'],    # 10-16時: 隣接は9時以前と17時以降
    'EVENING': ['AFTERNOON', 'NIGHT'],      # 17-20時: 隣接は16時以前と21時以降
    'NIGHT': ['EVENING', 'MORNING']         # 21-4時: 隣接は20時以前と5時以降
}


@dataclass
class Review:
    """口コミデータクラス"""
    reviewId: str
    spotName: str
    lat: float
    lon: float
    text: str
    weather: str
    timeSlot: str
    dayType: str
    likeCount: int
    viewCount: int
    createdAt: str
    photoKeys: List[str]
    userId: str
    status: str


@dataclass
class RecommendContext:
    """レコメンドコンテキストデータクラス"""
    lat: float
    lon: float
    weather: str
    timeSlot: str
    dayType: str


def calc_weather_score(review_weather: str, current_weather: str) -> float:
    """
    天気スコアを計算する (0.0〜1.0)
    
    Args:
        review_weather: 口コミの天気 (SUNNY/CLOUDY/RAINY/SNOWY/UNKNOWN)
        current_weather: 現在の天気 (SUNNY/CLOUDY/RAINY/SNOWY/UNKNOWN)
    
    Returns:
        float: 天気スコア
            - 一致: 1.0
            - UNKNOWN: 0.5 (中立)
            - 不一致: 0.0
    
    Validates: Requirements 4.1, 4.5
    Design: セクション6.2
    """
    if review_weather == 'UNKNOWN':
        return 0.5
    
    if review_weather == current_weather:
        return 1.0
    
    return 0.0


def calc_timeslot_score(review_slot: str, current_slot: str) -> float:
    """
    時間帯スコアを計算する (0.0〜1.0)
    
    Args:
        review_slot: 口コミの時間帯 (MORNING/AFTERNOON/EVENING/NIGHT)
        current_slot: 現在の時間帯 (MORNING/AFTERNOON/EVENING/NIGHT)
    
    Returns:
        float: 時間帯スコア
            - 一致: 1.0
            - 隣接: 0.5
            - 不一致: 0.0
    
    Validates: Requirements 4.1
    Design: セクション6.2
    """
    if review_slot == current_slot:
        return 1.0
    
    if review_slot in TIMESLOT_ADJACENT.get(current_slot, []):
        return 0.5
    
    return 0.0


def calc_distance_score(distance_m: float) -> float:
    """
    距離スコアを計算する (0.0〜1.0)
    
    Args:
        distance_m: 距離（メートル）
    
    Returns:
        float: 距離スコア
            - 計算式: max(0, 1 - d/5000)
            - 5km以上で0、0mで1.0
    
    Validates: Requirements 4.2
    Design: セクション6.2
    """
    return max(0.0, 1.0 - distance_m / 5000.0)


def calc_likes_score(like_count: int) -> float:
    """
    いいね数スコアを計算する (0.0〜1.0)
    
    Args:
        like_count: いいね数
    
    Returns:
        float: いいね数スコア
            - 計算式: min(1.0, like_count/100)
            - 100いいね以上で1.0
    
    Validates: Requirements 4.10
    Design: セクション6.2
    """
    return min(1.0, like_count / 100.0)


def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    2点間の距離をHaversine公式で計算する（メートル単位）
    
    Args:
        lat1: 地点1の緯度
        lon1: 地点1の経度
        lat2: 地点2の緯度
        lon2: 地点2の経度
    
    Returns:
        float: 距離（メートル）
    
    Design: セクション6.2
    """
    # 地球の半径（メートル）
    R = 6371000
    
    # 緯度経度をラジアンに変換
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)
    
    # Haversine公式
    a = math.sin(delta_phi / 2) ** 2 + \
        math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    distance = R * c
    return distance


def calc_total_score(
    weather_score: float,
    timeslot_score: float,
    distance_score: float,
    likes_score: float,
    weights: Optional[Dict[str, float]] = None
) -> float:
    """
    各スコアを重み付けして合計スコアを計算する
    
    Args:
        weather_score: 天気スコア (0.0〜1.0)
        timeslot_score: 時間帯スコア (0.0〜1.0)
        distance_score: 距離スコア (0.0〜1.0)
        likes_score: いいね数スコア (0.0〜1.0)
        weights: 重み設定（省略時はデフォルト）
    
    Returns:
        float: 合計スコア
    
    Validates: Requirements 4.3, 4.10
    Design: セクション6.1, 6.2
    """
    if weights is None:
        weights = DEFAULT_WEIGHTS
    
    total = (
        weights['weather'] * weather_score +
        weights['timeslot'] * timeslot_score +
        weights['distance'] * distance_score +
        weights['likes'] * likes_score
    )
    
    return total


def recommend(
    reviews: List[Review],
    context: RecommendContext,
    weights: Optional[Dict[str, float]] = None
) -> List[Review]:
    """
    コンテキストに基づいて口コミをレコメンド順にソートする
    
    Args:
        reviews: 口コミリスト
        context: レコメンドコンテキスト（現在位置・天気・時間帯）
        weights: 重み設定（省略時はデフォルト）
    
    Returns:
        List[Review]: 合計スコアの降順でソートされた口コミリスト
    
    Validates: Requirements 4.1, 4.2, 4.3, 4.8, 4.10
    Design: セクション6.1, 6.2, 6.3
    """
    if weights is None:
        weights = DEFAULT_WEIGHTS
    
    # 各口コミのスコアを計算
    scored_reviews = []
    
    for review in reviews:
        # 各スコア要素を計算
        w_score = calc_weather_score(review.weather, context.weather)
        t_score = calc_timeslot_score(review.timeSlot, context.timeSlot)
        
        # 距離を計算してスコア化
        distance = haversine_distance(
            context.lat, context.lon,
            review.lat, review.lon
        )
        d_score = calc_distance_score(distance)
        
        # いいね数スコア
        l_score = calc_likes_score(review.likeCount)
        
        # 合計スコアを計算
        total_score = calc_total_score(w_score, t_score, d_score, l_score, weights)
        
        scored_reviews.append((total_score, review))
    
    # スコアの降順でソート
    scored_reviews.sort(key=lambda x: x[0], reverse=True)
    
    # 口コミリストのみを返す
    return [review for _, review in scored_reviews]

File: src/utils/scoring.ts
// src/utils/scoring.ts
// スコアリング計算関数 — NagoTime コンテキスト対応レコメンド
// Requirements: 4.1〜4.10

import type { Review, Weather, TimeSlot, RecommendContext } from '../mocks/data/types'
import { haversine } from './geoUtils'

// ---- 重み定数（Requirements 4.10） ----
const WEIGHT_WEATHER = 0.30
const WEIGHT_TIME_SLOT = 0.25
const WEIGHT_DISTANCE = 0.30
const WEIGHT_LIKES = 0.15

// 距離スコアの正規化基準距離（メートル）
const DISTANCE_NORM_M = 5000

// いいねスコアの正規化上限
const LIKES_NORM_MAX = 100

// 時間帯の循環隣接順序
const TIME_SLOT_ORDER: TimeSlot[] = ['MORNING', 'AFTERNOON', 'EVENING', 'NIGHT']

/**
 * 天気スコアを計算する。
 * - 同一天気 → 1.0
 * - どちらかが UNKNOWN → 0.0
 * - それ以外 → 0.0
 *
 * Requirements: 4.3
 *
 * @param reviewWeather - 口コミ投稿時の天気
 * @param currentWeather - 現在の天気
 * @returns 0.0〜1.0
 */
export function calcWeatherScore(
  reviewWeather: Weather,
  currentWeather: Weather,
): number {
  if (reviewWeather === 'UNKNOWN' || currentWeather === 'UNKNOWN') {
    return 0.0
  }
  return reviewWeather === currentWeather ? 1.0 : 0.0
}

/**
 * 時間帯スコアを計算する。
 * - 同一時間帯 → 1.0
 * - 循環隣接（MORNING↔AFTERNOON、AFTERNOON↔EVENING、EVENING↔NIGHT、NIGHT↔MORNING）→ 0.5
 * - それ以外 → 0.0
 *
 * Requirements: 4.3
 *
 * @param reviewSlot - 口コミ投稿時の時間帯
 * @param currentSlot - 現在の時間帯
 * @returns 0.0〜1.0
 */
export function calcTimeSlotScore(
  reviewSlot: TimeSlot,
  currentSlot: TimeSlot,
): number {
  if (reviewSlot === currentSlot) {
    return 1.0
  }

  const len = TIME_SLOT_ORDER.length
  const reviewIdx = TIME_SLOT_ORDER.indexOf(reviewSlot)
  const currentIdx = TIME_SLOT_ORDER.indexOf(currentSlot)

  // 循環距離が 1 なら隣接
  const diff = Math.abs(reviewIdx - currentIdx)
  const circularDiff = Math.min(diff, len - diff)

  return circularDiff === 1 ? 0.5 : 0.0
}

/**
 * 距離スコアを計算する。
 * - `Math.max(0, 1 - distanceM / 5000)`
 * - 5000m 以上は 0.0、0m なら 1.0
 *
 * Requirements: 4.2, 4.3
 *
 * @param distanceM - ユーザーとスポット間の距離（メートル）
 * @returns 0.0〜1.0
 */
export function calcDistanceScore(distanceM: number): number {
  return Math.max(0, 1 - distanceM / DISTANCE_NORM_M)
}

/**
 * いいね数スコアを計算する。
 * - `Math.min(1.0, likeCount / 100)`
 * - 100いいね以上で最大値 1.0
 *
 * Requirements: 4.3
 *
 * @param likeCount - 口コミのいいね数
 * @returns 0.0〜1.0
 */
export function calcLikesScore(likeCount: number): number {
  return Math.min(1.0, likeCount / LIKES_NORM_MAX)
}

/**
 * 合計スコアを計算する。
 * - `0.30 * w + 0.25 * t + 0.30 * d + 0.15 * l`
 *
 * Requirements: 4.3, 4.10
 *
 * @param w - 天気スコア（0.0〜1.0）
 * @param t - 時間帯スコア（0.0〜1.0）
 * @param d - 距離スコア（0.0〜1.0）
 * @param l - いいねスコア（0.0〜1.0）
 * @returns 合計スコア（0.0〜1.0）
 */
export function calcTotalScore(
  w: number,
  t: number,
  d: number,
  l: number,
): number {
  return WEIGHT_WEATHER * w + WEIGHT_TIME_SLOT * t + WEIGHT_DISTANCE * d + WEIGHT_LIKES * l
}

/**
 * 口コミ一覧をコンテキストに基づいてスコアリングし、降順でソートして返す。
 *
 * 動作:
 * 1. PUBLISHED ステータスの口コミのみを対象にする（Requirements 4.8）
 * 2. 各口コミのスコアを計算（天気・時間帯・距離・いいね数）
 * 3. 合計スコアの降順でソートして返す（Requirements 4.3）
 *
 * @param reviews - 口コミ一覧
 * @param context - 現在のコンテキスト（位置情報・天気・時間帯）
 * @returns スコア降順にソートされた口コミ一覧（PUBLISHED のみ）
 */
export function recommend(reviews: Review[], context: RecommendContext): Review[] {
  const published = reviews.filter((r) => r.status === 'PUBLISHED')

  const scored = published.map((review) => {
    const distanceM = haversine(context.lat, context.lon, review.lat, review.lon)

    const w = calcWeatherScore(review.weather, context.weather)
    const t = calcTimeSlotScore(review.timeSlot, context.timeSlot)
    const d = calcDistanceScore(distanceM)
    const l = calcLikesScore(review.likeCount)
    const totalScore = calcTotalScore(w, t, d, l)

    return { review, totalScore }
  })

  scored.sort((a, b) => b.totalScore - a.totalScore)

  return scored.map((item) => item.review)
}
